File: src/adapters/engines/groupCooccurrenceEngine.ts
import type { RecommendationEngine } from '../../contracts/recommendationEngine'
import type { Recommendation, RecommendationRequest } from '../../domain/palette'
import { rgbDistance } from '../../lib/colorMath'

/** Explainable baseline engine for palettes with color-to-group training signals. */
export class GroupCooccurrenceEngine implements RecommendationEngine {
  readonly id = 'group-cooccurrence-v1'
  readonly name = 'Group co-occurrence'

  async recommend({ dataset, selectedColorIds, mode, limit }: RecommendationRequest): Promise<Recommendation[]> {
    if (selectedColorIds.length === 0) return []
    const selectedIds = new Set(selectedColorIds)
    const colorsById = new Map(dataset.colors.map((color) => [color.id, color]))
    const selected = selectedColorIds.flatMap((id) => {
      const color = colorsById.get(id)
      return color ? [color] : []
    })
    const observedGroups = new Set(selectedColorIds.flatMap((id) => dataset.groupsByColor[id] ?? []))

    const scored = dataset.colors.flatMap((color): Recommendation[] => {
      if (selectedIds.has(color.id)) return []
      const groups = dataset.groupsByColor[color.id] ?? []
      const shared = groups.reduce((count, groupId) => count + Number(observedGroups.has(groupId)), 0)
      if (shared === 0) return []
      const coOccurrence = shared / Math.sqrt(Math.max(1, groups.length * observedGroups.size))
      const distance = selected.reduce((sum, input) => sum + rgbDistance(color, input), 0) / selected.length
      const moodFactor = mode === 'quiet' ? 1 - distance : mode === 'vivid' ? distance : 1 - Math.abs(distance - 0.52)
      return [{ color, score: coOccurrence * 0.82 + moodFactor * 0.18, evidence: { label: 'shared', value: shared } }]
    })

    return scored.sort((left, right) => right.score - left.score).slice(0, limit)
  }
}

File: backend/adapters/group_cooccurrence_engine.py
from __future__ import annotations

import math
from typing import Optional

from ..domain import PaletteColor, PaletteDataset, Recommendation


class GroupCooccurrenceEngine:
    """Explainable baseline that learns from any palette's item-to-group matrix."""

    id = "group-cooccurrence-v1"
    name = "Group co-occurrence"

    def __init__(self):
        self._dataset: Optional[PaletteDataset] = None
        self._colors_by_id: dict[str, PaletteColor] = {}

    def fit(self, dataset: PaletteDataset) -> None:
        self._dataset = dataset
        self._colors_by_id = {color.id: color for color in dataset.colors}

    @staticmethod
    def _distance(left: PaletteColor, right: PaletteColor) -> float:
        return math.sqrt(sum((a - b) ** 2 for a, b in zip(left.rgb, right.rgb))) / 441.67

    def recommend(self, selected_color_ids: list[str], mode: str = "balanced", limit: int = 4) -> list[Recommendation]:
        if self._dataset is None:
            raise RuntimeError("Engine must be fit before inference")
        selected = [self._colors_by_id[color_id] for color_id in selected_color_ids if color_id in self._colors_by_id]
        if not selected:
            return []
        observed = frozenset().union(*(self._dataset.groups_by_color[color.id] for color in selected))
        scored: list[Recommendation] = []
        selected_ids = frozenset(color.id for color in selected)
        for candidate in self._dataset.colors:
            if candidate.id in selected_ids:
                continue
            groups = self._dataset.groups_by_color.get(candidate.id, frozenset())
            shared = len(groups & observed)
            if not shared:
                continue
            co_occurrence = shared / math.sqrt(max(1, len(groups) * len(observed)))
            distance = sum(self._distance(candidate, color) for color in selected) / len(selected)
            mood = 1 - distance if mode == "quiet" else distance if mode == "vivid" else 1 - abs(distance - .52)
            scored.append(Recommendation(candidate, co_occurrence * .82 + mood * .18, "shared", shared))
        return sorted(scored, key=lambda item: item.score, reverse=True)[:limit]
